src/components/CoSign/Cosign.js

```diff
@@ -0,0 +1,106 @@
+import React, {useContext} from 'react';
+import {Route, Routes, useNavigate} from 'react-router-dom';
+import {useImmerReducer} from 'use-immer';
+
+import {destroy} from 'api';
+import ErrorBoundary from 'components/ErrorBoundary';
+import {CosignSummary} from 'components/Summary';
+
+import {ConfigContext} from '../../Context';
+import CosignDone from './CosignDone';
+
+const initialState = {
+  submission: null,
+  privacyInfo: {
+    requiresPrivacyConsent: true,
+    privacyLabel: '',
+    policyAccepted: false,
+  },
+  summaryData: [],
+  reportUrl: '',
+  cosignedSubmission: null,
+};
+
+const reducer = (draft, action) => {
+  switch (action.type) {
+    case 'SUBMISSION_LOADED': {
+      draft.submission = action.payload;
+      break;
+    }
+    case 'LOADED_PRIVACY_INFO': {
+      draft.privacyInfo = {...draft.privacyInfo, ...action.payload};
+      break;
+    }
+    case 'LOADED_SUMMARY_DATA': {
+      draft.summaryData = action.payload;
+      break;
+    }
+    case 'TOGGLE_PRIVACY_CHECKBOX': {
+      draft.privacyInfo.policyAccepted = !draft.privacyInfo.policyAccepted;
+      break;
+    }
+    case 'COSIGN_COMPLETE': {
+      return {...initialState, reportUrl: action.payload};
+    }
+    case 'RESET': {
+      return initialState;
+    }
+    default: {
+      throw new Error(`Unknown action ${action.type}`);
+    }
+  }
+};
+
+const Cosign = ({...props}) => {
+  const [state, dispatch] = useImmerReducer(reducer, initialState);
+  const navigate = useNavigate();
+  const config = useContext(ConfigContext);
+
+  const onDataLoaded = data => {
+    dispatch({type: 'LOADED_PRIVACY_INFO', payload: data.privacyInfo});
+    dispatch({type: 'LOADED_SUMMARY_DATA', payload: data.summaryData});
```

Comment:
it looks like these can be merged into a single reducer action, so that you only have a single `dispatch` call?

I'm not sure how much it matters though, I believe with React 18 there is now more aggressive batching even for non-event callbacks. See https://github.com/reactwg/react-18/discussions/21